pkg/restore/restore.go

```diff
@@ -404,104 +355,66 @@ func (ctx *context) execute() (Result, Result) {
 	// need to set this for additionalItems to be restored
 	ctx.restoreDir = dir
 
+	var (
+		existingNamespaces = sets.NewString()
+		processedResources = sets.NewString()
+	)
+
 	backupResources, err := archive.NewParser(ctx.log, ctx.fileSystem).Parse(ctx.restoreDir)
 	if err != nil {
 		errs.AddVeleroError(errors.Wrap(err, "error parsing backup contents"))
 		return warnings, errs
 	}
 
-	existingNamespaces := sets.NewString()
-
-	for _, resource := range ctx.prioritizedResources {
-		// we don't want to explicitly restore namespace API objs because we'll handle
-		// them as a special case prior to restoring anything into them
-		if resource == kuberesource.Namespaces {
+	// Iterate through an ordered list of resources to restore, checking each one to see if it should be restored.
+	// Note that resources *may* be in this list twice, i.e. once due to being a prioritized resource, and once due
+	// to being in the backup tarball. We can't de-dupe this upfront, because it's possible that items in the prioritized
+	// resources list may not be fully resolved group-resource strings (e.g. may be specfied as "po" instead of "pods"),
+	// and we don't want to fully resolve them via discovery until we reach them in the loop, because it is possible
+	// that the resource/API itself is being restored via a custom resource definition, meaning it's not available via
+	// discovery prior to beginning the restore.
+	//
+	// Since we keep track of the fully-resolved group-resources that we *have* restored, we won't try to restore a
+	// resource twice even if it's in the ordered list twice.
+	for _, resource := range getOrderedResources(ctx.resourcePriorities, backupResources) {
+		// try to resolve the resource via discovery to a complete group/version/resource
+		gvr, _, err := ctx.discoveryHelper.ResourceFor(schema.ParseGroupResource(resource).WithVersion(""))
+		if err != nil {
+			// TODO should we refresh discovery and try again?
+			ctx.log.WithField("resource", resource).Infof("Skipping restore of resource because it cannot be resolved via discovery")
 			continue
 		}
+		groupResource := gvr.GroupResource()
 
-		resourceList := backupResources[resource.String()]
-		if resourceList == nil {
+		// check if we've already restored this resource (this would happen if the restore
+		// we're currently looking at was already restored because it was a prioritized
+		// resource, and now we're looking at it as part of the backup contents).
+		if processedResources.Has(groupResource.String()) {
+			ctx.log.WithField("resource", groupResource.String()).Infof("Skipping restore of resource because it's already been processed")
```

Comment:
Should this be info or debug? Based on your testing, how often do we see this line?

Reply:
probably makes sense to bump down to debug, otherwise it'll come up regularly for any prioritized resource